### ISSUE TITLE
Unable to segment a TEM image using a script

### ISSUE THREAD

**Issue by Stoyan-I-A** [neuropoly/axondeepseg#245]
I tried to use the FSLeyes plugin to segment the image found in [OSF](https://osf.io/rtbwc/) under `test_segmentation/test_tem_image/image1_tem/image.png `

![image](https://user-images.githubusercontent.com/43790661/62840596-e0b76900-bc51-11e9-89c8-b94cd2df9ebb.png)
[pixel_size_in_micrometer.txt](https://github.com/neuropoly/axondeepseg/files/3490531/pixel_size_in_micrometer.txt)

But it returned me an error when I tried to apply the TEM model.

I wrote this script to replicate the instructions used in the plugin : 
```
import os
import json

import AxonDeepSeg
from AxonDeepSeg.apply_model import axon_segmentation

image_directory = "/home/stoyan/Desktop/test_segmentation/test_tem_image/image1_tem"
image_name = "image.png"
image_path = image_directory + "/" + image_name

model_path = os.path.dirname(AxonDeepSeg.__file__)
model_path = model_path = os.path.join(
                model_path, "models", "default_TEM_model_v1"
             )
model_configfile = os.path.join(model_path, "config_network.json")
with open(model_configfile, "r") as fd:
    config_network = json.loads(fd.read())

prediction = axon_segmentation(
            [image_directory],
            [image_name],
            model_path,
            config_network,
            verbosity_level=3,
            )
```

Here is the output of the console :
```
Loading acquisitions ...
Rescaling acquisitions to the target resolution ...
Graph construction ...
Beginning inference ...
processing patch 1 of 1
Traceback (most recent call last):
  File "/home/stoyan/PycharmProjects/ADS_dev2/TEM_test.py", line 25, in <module>
    verbosity_level=3,
  File "/home/stoyan/axondeepseg/AxonDeepSeg/apply_model.py", line 277, in axon_segmentation
    verbosity_level=verbosity_level)
  File "/home/stoyan/axondeepseg/AxonDeepSeg/apply_model.py", line 130, in apply_convnet
    prediction_proba_activate=prediction_proba_activate)
  File "/home/stoyan/axondeepseg/AxonDeepSeg/apply_model.py", line 481, in perform_batch_inference
    p = model.predict(batch_x)
  File "/home/stoyan/miniconda/envs/ADS_dev2/lib/python3.6/site-packages/keras/engine/training.py", line 1149, in predict
    x, _, _ = self._standardize_user_data(x)
  File "/home/stoyan/miniconda/envs/ADS_dev2/lib/python3.6/site-packages/keras/engine/training.py", line 751, in _standardize_user_data
    exception_prefix='input')
  File "/home/stoyan/miniconda/envs/ADS_dev2/lib/python3.6/site-packages/keras/engine/training_utils.py", line 138, in standardize_input_data
    str(data_shape))
ValueError: Error when checking input: expected input_1 to have shape (512, 512, 3) but got array with shape (62, 94, 3)

Process finished with exit code 1
```
However, I am able to segment the image using the terminal command line with no problem. ( This was fixed in #230 )
Could there be something wrong with the script that I am using?

Edit: I should also note that I don't have this problem with SEM images.

**Comment by mathieuboudreau** [comment]
Is it possible that you're not using the correct pixel size? Sounds like it expected a patch of 512x512, but got 62x94 when it was processing it. Make sure it didn't use one of the default versions maybe?

**Comment by mathieuboudreau** [comment]
Or change the resampled pixel size so that it's able to resample 512x512 patches?

**Comment by Stoyan-I-A** [comment]
> Is it possible that you're not using the correct pixel size?

I doubt this is causing the problem since I was able to segment the image using the terminal without specifying another pixel size.
I will still try to modify the pixel size to see if it fixes the issue.

**Comment by Stoyan-I-A** [comment]
I tried changing the pixel size in micrometer to 0.01 (the original pixel size of the image was 0.0025), the last line of the output error became `ValueError: Error when checking input: expected input_1 to have shape (512, 512, 3) but got array with shape (249, 136, 3)`

**Comment by mathieuboudreau** [comment]
I don't see the way you're passing along the resampled resolution and/or acquired resolution to `axon_segmentation()` in your plugin:

https://github.com/neuropoly/axondeepseg/blob/dfdd57f0dd1e9eb728ac436f0f355efd723c216b/ads_plugin.py#L284-L290

If you don't pass the argument `resampled_resolutions`, by default it uses 0.1:

https://github.com/neuropoly/axondeepseg/blob/dfdd57f0dd1e9eb728ac436f0f355efd723c216b/AxonDeepSeg/apply_model.py#L197-L201

**Comment by Stoyan-I-A** [comment]
I tried to change the `resampled_resolution` to 0.01. The output no longer returned an error, but the results are not what they should be :
![AxonDeepSeg](https://user-images.githubusercontent.com/43790661/62902621-05284980-bd15-11e9-8e24-edb5627ebff5.png)
I tried other values of `resampled_resolution` and I also tried changing the `acquired_resolution` but the results were similar.

In contrast, this is what I obtained using the command line :
![image_seg-axonmyelin](https://user-images.githubusercontent.com/43790661/62902706-47518b00-bd15-11e9-88e3-a77da65850a9.png)

**Comment by jcohenadad** [comment]
Weird that the few segmented axons on the top image don’t correspond to any axons at the bottom. Are you sure you are inputting the correct image? Without any transformation to it?

**Comment by mathieuboudreau** [comment]
If the command line is working well, but the plugin isn't, then it can only be due to a difference in how the segmentation is called in your plugin (either not all the same option, or some values are different).

For reference, here is where the command line calls ADS to perform the segmentation:

https://github.com/neuropoly/axondeepseg/blob/785012eb16ab7d8d9c3a2526dc0edf82d35f8ba4/AxonDeepSeg/segment.py#L381-L385

which in turn, calls `axon_segmentation()` here:

https://github.com/neuropoly/axondeepseg/blob/785012eb16ab7d8d9c3a2526dc0edf82d35f8ba4/AxonDeepSeg/segment.py#L86-L94

So I think a good first start would be to debug all the `axon_segmentation()` function arguments for both cases (e.g. at the top of `axon_segmentation()` in [this module](https://github.com/neuropoly/axondeepseg/blob/785012eb16ab7d8d9c3a2526dc0edf82d35f8ba4/AxonDeepSeg/apply_model.py#L197-L224), print out all the parameter values)

**Comment by mathieuboudreau** [comment]
@Stoyan-I-A I just tried debugging this - two things to note.

1 - I encountered the bug referenced in #238, so I assume that to get around this you applied the proposed fix but didn't push it to GitHub, correct? (I had to do that to get something segmented).

2 - I ran the segmentation on your image two ways & printed out the `axon_segmentation()` inputs; using your FSLeyes plugin (with TEM selected as the modality, and using the command line interface (with this command: `axondeepseg -t TEM -i image.png -v 3`). Here is a comparison of the outputs (left is command line, right is FSLeyes):

<img width="1001" alt="Capture d’écran 2019-08-13 à 15 03 44" src="https://user-images.githubusercontent.com/1421029/62963691-112b0e80-bddc-11e9-9773-c224a9f5de02.png">

So, I see a couple of bugs in your plugin. 1) You aren't actually passing the acquired resolution (i.e. what is in the pixel_size_in_micrometer.txt file), and 2) you aren't passing/handling the TEM option(s) correctly when it is selected by the GUI (I can see this because the default resample_resolution for SEM is 0.1 and TEM is 0.01, and you are using 0.1 when you shouldn't be from the image). 

These two sections of the command line function might help you:

https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/segment.py#L242-L258

https://github.com/neuropoly/axondeepseg/blob/a7da6a1e7bf56a84b917045b9cdee1b8d0d48246/AxonDeepSeg/segment.py#L320-L322

Where you would need to pass `resolution_model` to the `resample_resolution` parameter of `axon_segmentation()`.

**Comment by Stoyan-I-A** [comment]
I didn't fix the bug in #238 since it wasn't causing any problem (so far). Since it is a known issue, I simply removed the output lines related to it on the top post.

Thank you for looking into this. I will make modifications to the script and update this issue if I make any progress or encounter problems.

**Comment by Stoyan-I-A** [comment]
Okay, I managed to resolve the issue. I just need to make modifications to the plugin script. 
Here's what I did:
- I set the `resampled_resolution` to 0.01 and the `aquired_resolution` to 0.0025.
- I tried to call `axon_segmentation()` with these parameters. However, I still obtained bad results.
- Instead of calling `axon_segmentation()`, I called `segment_image()` and it returned the same results as the command line interface.

I will close this issue once I push the changes.

**Comment by mathieuboudreau** [comment]
Great!

What's weird is that from what I see, the only difference that `segment_image()` does is invert the brightness (which nonetheless is a major change). 

https://github.com/neuropoly/axondeepseg/blob/ebf74c1a928ef946304d821e1e5abf0b692fd440/AxonDeepSeg/segment.py#L167-L170

@vs74 why do you think that is? Is the TEM training done with inverted brightness too? I wouldn't think it would have to be.

**Comment by Stoyan-I-A** [comment]
Fixed in db4cf81abaefb35d8d7c645617e89e154f6cbd9d
